### PR TITLE
RTN24, RTN21 and partial RTN4h

### PR DESCRIPTION
**Pull request by withakay**
`(RTN24) A connected client may receive a CONNECTED ProtocolMessage from Ably at any point (though is typically triggered by a reauth, see RTC8a). The connectionDetails in the ProtocolMessage must override any stored details, see RTN21. The Connection should emit an UPDATE event with a ChannelStateChange object, which should have both previous and current attributes set to CONNECTED, and the reason attribute set to to the error member of the CONNECTED ProtocolMessage (if any). (Note that UPDATE should be the only event emitted: in particular, the library must not emit an CONNECTED event if the client was already connected, see RTN4h).	`		

`(RTN21) If the CONNECTED ProtocolMessage contains a connectionDetails property, the attributes within ConnectionDetails will be used as the defaults for this client library, overriding any configured options at the time the CONNECTED ProtocolMessage is received`

`(RTN4h) It can emit an UPDATE event (this is the only ConnectionEvent which is not a ConnectionState). This is used for changes to connection conditions for which the ConnectionState (e.g. CONNECTED) does not change. (The library must never emit a ConnectionState event for a state equal to the previous state).`

The final part of RTN4h `The library must never emit a ConnectionState event for a state equal to the previous state` is not covered by these tests. 

There is a [pre-existing test for RTN8b](https://github.com/ably/ably-dotnet/blob/e76f6c94df952171a38bc84a2f5cfd1ac1a42e6f/src/IO.Ably.Tests/Realtime/ConnectionStateTests/ConnectedStateSpecs.cs#L119) that additionally covers part of RTN21 at a lower level.

### CODE REVIEW

File: src/IO.Ably.Shared/Transport/ConnectionManager.cs [hunk 1]
```diff
@@ -116,30 +116,35 @@ public Task SetState(ConnectionStateBase newState, bool skipAttach = false)
...
                     lock (_stateSyncLock)
                     {
-                        if (State.State == newState.State)
+                        if (!newState.IsUpdate)
                         {
-                            if (Logger.IsDebug)
+                            if (State.State == newState.State)
                             {
-                                Logger.Debug($"xx State is already {State.State}. Skipping SetState action.");
+                                if (Logger.IsDebug)
```
Comment (funkyboy) [inline src/IO.Ably.Shared/Transport/ConnectionManager.cs]: This might become quite tedious to write. Have you considered just writing `Logger.Debug( ... )` and move the `if` check within `Logger.Debug( ... )`

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs [hunk 2]
```diff
@@ -814,6 +814,61 @@ public async Task
...
+        [Trait("spec", "RTN24")]
+        [Trait("spec", "RTN21")]
+        [Trait("spec", "RTN4h")]
+        public async Task WhenConnectedMessageReceived_ShouldEmitUpdate(Protocol protocol)
+        {
+            var updateAwaiter = new TaskCompletionAwaiter(5000);
+            var client = await GetRealtimeClient(protocol, (options, settings) =>
+            {
+                options.UseTokenAuth = true;
+                options.AutoConnect = true;
```
Comment (funkyboy) [inline src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: This shouldn't be needed

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs [hunk 3]
```diff
@@ -814,6 +814,61 @@ public async Task
...
+            });
+
+            await client.WaitForState(ConnectionState.Connected);
+
+            client.Connection.ConnectionStateTtl.Should().NotBe(TimeSpan.MaxValue);
```
Comment (funkyboy) [inline src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: Why is this relevant to test?

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs [hunk 4]
```diff
@@ -814,6 +814,61 @@ public async Task
...
+            var key = client.Connection.Key;
+
+            client.Connection.Once(state =>
+            {
+                // RTN4h - can emit UPDATE event
+                if (state.Event == ConnectionEvent.Update)
+                {
+                    // should have both previous and current attributes set to CONNECTED
+                    state.Current.Should().Be(ConnectionState.Connected);
+                    state.Previous.Should().Be(ConnectionState.Connected);
```
Comment (funkyboy) [inline src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: I'd add also a check for the error field

File: src/IO.Ably.Shared/Transport/ConnectionManager.cs (hunk 1 above)
Comment (withakay) [reply src/IO.Ably.Shared/Transport/ConnectionManager.cs]: I have, and this is a bit of an oddity, technically it is not needed but the code base already uses this method a lot so I have just been trying to be consistent for now. IT could be something that we clean up on the project level when higher priority things are out the way...?

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs (hunk 3 above)
Comment (withakay) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: As per RTN21 `ConnectionStateTtl` should be set to `TimeSpan.MaxValue` via the `ConnectionDetails` later (line 852), this shows that it didn't start out at that value.

File: src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs (hunk 4 above)
Comment (withakay) [reply src/IO.Ably.Tests.Shared/Realtime/ConnectionSandBoxSpecs.cs]: Good spot, I will add a check for that.

File: src/IO.Ably.Shared/Transport/ConnectionManager.cs (hunk 1 above)
Comment (funkyboy) [reply src/IO.Ably.Shared/Transport/ConnectionManager.cs]: Sure. Can you open an issue about it so we don’t forget?

Comment (withakay) [reply src/IO.Ably.Shared/Transport/ConnectionManager.cs]: Added issue #229 to cover this.
